external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java

```diff
@@ -0,0 +1,547 @@
+        }
+    }
+
+    private void setAcked(TopicPartition tp, long fetchOffset) {
+        // If this partition was previously assigned to this spout, leave the acked offsets as they were to resume where it left off
+        if (!consumerAutoCommitMode && !acked.containsKey(tp)) {
+            acked.put(tp, new OffsetEntry(tp, fetchOffset));
+        }
+    }
+
+    // ======== Next Tuple =======
+
+    @Override
+    public void nextTuple() {
+        if (initialized) {
+            if (commit()) {
+                commitOffsetsForAckedTuples();
+            }
+
+            if (poll()) {
+                setWaitingToEmit(pollKafkaBroker());
+            }
+
+            if (waitingToEmit()) {
+                emit();
+            }
+        } else {
+            LOG.debug("Spout not initialized. Not sending tuples until initialization completes");
+        }
+    }
+
+    private boolean commit() {
+        return !consumerAutoCommitMode && commitTimer.isExpiredResetOnTrue();    // timer != null for non auto commit mode
+    }
+
+    private boolean poll() {
+        return !waitingToEmit() && numUncommittedOffsets < kafkaSpoutConfig.getMaxUncommittedOffsets();
+    }
+
+    private boolean waitingToEmit() {
+        return waitingToEmit != null && waitingToEmit.hasNext();
+    }
+
+    public void setWaitingToEmit(ConsumerRecords<K,V> consumerRecords) {
+        List<ConsumerRecord<K,V>> waitingToEmitList = new LinkedList<>();
+        for (TopicPartition tp : consumerRecords.partitions()) {
+            waitingToEmitList.addAll(consumerRecords.records(tp));
+        }
+        waitingToEmit = waitingToEmitList.iterator();
+        LOG.trace("Records waiting to be emitted {}", waitingToEmitList);
+    }
+
+    // ======== poll =========
+    private ConsumerRecords<K, V> pollKafkaBroker() {
+        doSeekRetriableTopicPartitions();
+
+        final ConsumerRecords<K, V> consumerRecords = kafkaConsumer.poll(kafkaSpoutConfig.getPollTimeoutMs());
+        final int numPolledRecords = consumerRecords.count();
+        LOG.debug("Polled [{}] records from Kafka. NumUncommittedOffsets=[{}]", numPolledRecords, numUncommittedOffsets);
+        return consumerRecords;
+    }
+
+    private void doSeekRetriableTopicPartitions() {
+        final Set<TopicPartition> retriableTopicPartitions = retryService.retriableTopicPartitions();
+
+        for (TopicPartition rtp : retriableTopicPartitions) {
+            final OffsetAndMetadata offsetAndMeta = acked.get(rtp).findNextCommitOffset();
+            if (offsetAndMeta != null) {
+                kafkaConsumer.seek(rtp, offsetAndMeta.offset() + 1);  // seek to the next offset that is ready to commit in next commit cycle
+            } else {
+                kafkaConsumer.seekToEnd(rtp);    // Seek to last committed offset
+            }
+        }
+    }
+
+    // ======== emit  =========
+    private void emit() {
+        emitTupleIfNotEmitted(waitingToEmit.next());
+        waitingToEmit.remove();
```

Comment:
I want to add that I am still +1 on this, even without this change.  This is minor.